Add automated tooling and a guide for building and installing Kedro on a Databricks cluster

jmholzer opened the pull request:
## Description
Resolves #2318.

This PR adds tooling for building a local version of Kedro and uploading it to the developer's Databricks cluster. This greatly speeds up development time and is an important step towards improving Kedro for Databricks.

It also adds documentation for using this tooling.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

astrojuanlu commented:
🤖 Rendered version: https://kedro--2369.org.readthedocs.build/en/2369/contribution/development_for_databricks.html

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+## Introduction
+
+This page describes how to efficiently develop features and fixes for Kedro on Databricks. Using this guide, you will be able to quickly test your locally modified version of Kedro on Databricks as part of a build-and-test development cycle.
```
@jmholzer I was initially unsure why this is part of the contribution documentation rather than a revision of the deployment page we already have. I think it's because it's for people who are working on Kedro code itself (not a Kedro project) and need to test the updated Kedro version -- is that correct? I would bring that out a whole lot more and point to the deployment page to say "If you just wanna test your Kedro project, this isn't the page for you, go here instead" kind of thing. Likewise, i'd add something to the existing deployment page to push Kedro contributors over to this one.

jmholzer replied:
Correct, this is only for developers looking to improve Kedro 🙂 you're right I should have made this more clear.

I've added the notes you suggested on both the project deployment guide and on this guide (in [64e3afb](https://github.com/kedro-org/kedro/pull/2369/commits/64e3afbc491019c9e962d240ab69984ad0d21910) and [f5e6018](https://github.com/kedro-org/kedro/pull/2369/commits/f5e6018ce0bc66e8c4924e7cb0f9c167d5394c70)).

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+## Prerequisites
+
+You will need the following prerequisites to follow this guide:
```
```suggestion
You will need the following to follow this guide:
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+You will need the following prerequisites to follow this guide:
+
+1. Python **version >=3.8** installed.
```
Unless it's a sequence of steps (e.g. in a tutorial) you don't need to number the bullets but can leave them as symbolic.

stichbury replied:
```suggestion
1. Python **version >=3.8**.
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+You will need the following prerequisites to follow this guide:
+
+1. Python **version >=3.8** installed.
+2. A Python virtual environment activated in which the [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html) is installed with [authentication for your workspace](https://docs.databricks.com/dev-tools/cli/index.html#set-up-the-cli).
```
```suggestion
2. An activated Python virtual environment into which you have installed the [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html) with [authentication for your workspace](https://docs.databricks.com/dev-tools/cli/index.html#set-up-the-cli).
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+1. Python **version >=3.8** installed.
+2. A Python virtual environment activated in which the [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html) is installed with [authentication for your workspace](https://docs.databricks.com/dev-tools/cli/index.html#set-up-the-cli).
+3. Access to a Databricks workspace with an [existing cluster](https://docs.databricks.com/clusters/create-cluster.html).
```
```suggestion
3. Access to a Databricks workspace on an [existing cluster](https://docs.databricks.com/clusters/create-cluster.html).
```

stichbury replied:
No mention of type of cluster, e.g. AWS? We have a deployment guide and it seems to only support AWS at present, so is that a prerequisite or can any cloud provider cluster be used? Probably worth saying or pointing to a relevant databricks link?

jmholzer replied:
This is a good point, though creating a cluster on Databricks is actually a cloud-agnostic process on Databricks. The linked documentation actually covers the topic really well and the user can switch the cloud provider in the top right to view the (very minor) differences for their specific cloud. For this reason I don't think it's important that we specify a cloud here.

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+1. Python **version >=3.8** installed.
+2. A Python virtual environment activated in which the [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html) is installed with [authentication for your workspace](https://docs.databricks.com/dev-tools/cli/index.html#set-up-the-cli).
+3. Access to a Databricks workspace with an [existing cluster](https://docs.databricks.com/clusters/create-cluster.html).
+4. [GNU `make`](https://www.gnu.org/software/make/) installed.
```
```suggestion
4. [GNU `make`](https://www.gnu.org/software/make/).
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+2. A Python virtual environment activated in which the [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html) is installed with [authentication for your workspace](https://docs.databricks.com/dev-tools/cli/index.html#set-up-the-cli).
+3. Access to a Databricks workspace with an [existing cluster](https://docs.databricks.com/clusters/create-cluster.html).
+4. [GNU `make`](https://www.gnu.org/software/make/) installed.
+5. [`git` installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git).
```
```suggestion
5. [`git`](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git).
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+3. Access to a Databricks workspace with an [existing cluster](https://docs.databricks.com/clusters/create-cluster.html).
+4. [GNU `make`](https://www.gnu.org/software/make/) installed.
+5. [`git` installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git).
+6. A local clone of the [Kedro](https://github.com/kedro-org/kedro) repository for you to edit.
```
```suggestion
6. A local clone of [the Kedro GitHub repository](https://github.com/kedro-org/kedro).
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+This page describes how to efficiently develop features and fixes for Kedro on Databricks. Using this guide, you will be able to quickly test your locally modified version of Kedro on Databricks as part of a build-and-test development cycle.
+
+## Prerequisites
```
You've not made any mention of different OS -- are the instructions at all different depending on whether you're on Windows or MacOS? For example, installing GNUMake or Git?

jmholzer replied:
All requirements are available on all major operating systems.

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+# Developing Kedro for Databricks
```
```suggestion
# How to deploy a development version of Kedro to Databricks
```

stichbury replied:
That's just a preliminary suggestion, so feel free to throw some other headings around! I'd prefer not to use gerunds in section headings as a style choice, but also, I think in this case, the heading isn't particularly saying what the page is for/about (see my comment below). So making it super-clear (and removing the '-ing' would be awesome).

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+5. [`git` installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git).
+6. A local clone of the [Kedro](https://github.com/kedro-org/kedro) repository for you to edit.
+
+## Installing your Kedro builds on Databricks
```
```suggestion
## How to install a build of Kedro onto Databricks
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+## Installing your Kedro builds on Databricks
+
+The development workflow for Kedro on Databricks is similar to the one for Kedro in general. The main difference comes when manually testing your changes on Databricks. You can develop and test your changes locally, but you will need to build and deploy the wheel file to Databricks to test it on a cluster.
```
```suggestion
The development workflow for Kedro on Databricks is similar to the one for Kedro in general, when you develop and test your changes locally. The main difference comes when manually testing your changes on Databricks, since you will need to build and deploy the wheel file to Databricks to test it on a cluster.
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+The development workflow for Kedro on Databricks is similar to the one for Kedro in general. The main difference comes when manually testing your changes on Databricks. You can develop and test your changes locally, but you will need to build and deploy the wheel file to Databricks to test it on a cluster.
+
+To make developing Kedro for Databricks easier, Kedro comes with a `Makefile` target named `databricks-build` that automates the process of building a wheel file and installing this on your Databricks cluster. This is the recommended way of developing for Databricks because of the amount of time it saves.
```
```suggestion
To make developing Kedro for Databricks easier, Kedro comes with a `Makefile` target named `databricks-build` that automates the process of building a wheel file and installing this on your Databricks cluste to save development time.
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+To make developing Kedro for Databricks easier, Kedro comes with a `Makefile` target named `databricks-build` that automates the process of building a wheel file and installing this on your Databricks cluster. This is the recommended way of developing for Databricks because of the amount of time it saves.
+
+### Setup
```
```suggestion
### Setup the Databricks CLI to test a Kedro build
```

stichbury replied:
I've suggested this so it's more detailed (for organic search) and easier to spot in the table of contents.

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+### Setup
+
+In order to use `databricks-build`, you must have completed the [setup of the Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html#set-up-the-cli).
```
```suggestion
Before you use `make databricks-build`, you must [set up the Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html#set-up-the-cli).
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+In order to use `databricks-build`, you must have completed the [setup of the Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html#set-up-the-cli).
+
+The only remaining step is to create and environment variable with the ID of the cluster you are using to test your Kedro build. You can find the ID by executing the Databricks CLI command `databricks cluster list` and looking for the name of your chosen cluster.
```
```suggestion
Next, create an environment variable with the ID of the cluster you are using to test your Kedro build. You can find that ID by executing the Databricks CLI command `databricks cluster list` and looking for the name of your chosen cluster.
```

jmholzer replied:
You're right about this. I have added an example of the output of the command in [9056ef4](https://github.com/kedro-org/kedro/pull/2369/commits/9056ef421837c947c20126bd4f1f645a8b55e18a).

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+The only remaining step is to create and environment variable with the ID of the cluster you are using to test your Kedro build. You can find the ID by executing the Databricks CLI command `databricks cluster list` and looking for the name of your chosen cluster.
+
+Once you have found the cluster ID, you must export it to an environment variable named `DATABRICKS_CLUSTER_ID`, on Linux or macOS:
```
What about Windows?

stichbury commented on docs/source/contribution/development_for_databricks.md:
````diff
@@ -0,0 +1,73 @@
+export DATABRICKS_CLUSTER_ID=<your-cluster-id>
+```
+
+### Using `make databricks-build`
````
```suggestion
### How to use `make databricks-build` to test your Kedro build
```

stichbury replied:
Just a suggestion, feel free to modify

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+### Using `make databricks-build`
+
+With the setup complete, you can use `databricks-build`. Navigate to the parent directory of your Kedro development repository and run:
```
```suggestion
With the setup complete, you can use `make databricks-build`. Navigate to the parent directory of your Kedro development repository and run:
```

stichbury replied:
This is locally, right, in your terminal. Maybe worth saying "In your console, navigate to..."

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+4. Queues your updated wheel file for installation
+5. Restarts your cluster to apply the changes.
+
+Note that your cluster will be unavailable while it is restarting. You can poll the status of your cluster using the Databricks CLI command `databricks clusters get --cluster-id <your-cluster-id> | grep state` (macOS, Linux).
```
```suggestion
Note that your cluster will be unavailable while it restarts. You can poll the status of the cluster using the Databricks CLI command `databricks clusters get --cluster-id <your-cluster-id> | grep state` (macOS, Linux).
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,73 @@
+
+Note that your cluster will be unavailable while it is restarting. You can poll the status of your cluster using the Databricks CLI command `databricks clusters get --cluster-id <your-cluster-id> | grep state` (macOS, Linux).
+
+After waiting for your cluster to restart, you should verify that your modified version of Kedro has installed correctly. To do this, run `databricks libraries list --cluster-id <your-cluster-id>`. If installation was successful, you should see the following output:
```
```suggestion
Once the cluster has restarted, you should verify that your modified version of Kedro has been correctly installed. Run `databricks libraries list --cluster-id <your-cluster-id>`. If installation was successful, you should see the following output:
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
````diff
@@ -0,0 +1,73 @@
+}
+```
+
+Any runs of a Kedro project on this cluster will now reflect your latest local changes to Kedro.
````
Maybe worth saying you can then test your changes or run a Kedro project or whatever, to make it clear they're all done and can now proceed.

stichbury reviewed (changes requested):
Thanks @jmholzer this is really good. I made a fair few comments because I think this is the kind of page that needs crystal clarity, but really it was almost there anyway, so apologies if it looks a bit mangled now.

jmholzer commented:
> Thanks @jmholzer this is really good. I made a fair few comments because I think this is the kind of page that needs crystal clarity, but really it was almost there anyway, so apologies if it looks a bit mangled now.

Awesome, thanks so much for the feedback @stichbury. I have addressed the comments you left in my latest round of commits.

stichbury commented on docs/source/contribution/development_for_databricks.md:
````diff
@@ -0,0 +1,89 @@
+## Introduction
+
+```{note}
+This page is for people developing Kedro, users working on their own Kedro projects should see the documentation for [deploying Kedro projects on Databricks](../deployment/databricks.md).
````
```suggestion
This page is for people developing changes to Kedro that need to test them on Databricks. If you are working on a Kedro project and need more information about project-deployment, consult the [documentation for deploying Kedro projects on Databricks](../deployment/databricks.md).
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
````diff
@@ -0,0 +1,89 @@
+This page is for people developing Kedro, users working on their own Kedro projects should see the documentation for [deploying Kedro projects on Databricks](../deployment/databricks.md).
+```
+
+This guide describes how to efficiently develop features and fixes for Kedro on Databricks. Using this guide, you will be able to quickly test your locally modified version of Kedro on Databricks as part of a build-and-test development cycle.
````
I'd be tempted to move these two sentences to precede the note rather than follow it, but will leave it up to you.

jmholzer replied:
That's fair, it's probably bad form to start a section with a note. Fixed this.

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,89 @@
+
+Before you use `make databricks-build`, you must [set up the Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html#set-up-the-cli).
+
+The only remaining step is to create and environment variable with the ID of the cluster you are using to test your Kedro build. You can find the ID by executing the Databricks CLI command `databricks clusters list` and looking for the Cluster ID to the left of the name of your chosen cluster, for instance:
```
```suggestion
Next, create and environment variable with the ID of the cluster you are using to test your Kedro build. You can find the ID by executing the Databricks CLI command `databricks clusters list` and looking for the Cluster ID to the left of the name of your chosen cluster, for instance:
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,89 @@
+
+### How to use `make databricks-build` to test your Kedro build
+
+With the setup complete, you can use `databricks-build`. In your terminal, navigate to the parent directory of your Kedro development repository and run:
```
```suggestion
With the setup complete, you can use `make databricks-build`. In your terminal, navigate to the parent directory of your Kedro development repository and run:
```

stichbury commented on docs/source/contribution/development_for_databricks.md:
```diff
@@ -0,0 +1,89 @@
+4. Queues your updated wheel file for installation
+5. Restarts your cluster to apply the changes.
+
+Note that your cluster will be unavailable while it restarts. You can poll the status of the cluster using the Databricks CLI command `databricks clusters get --cluster-id <your-cluster-id> | grep state` (macOS, Linux).
```
Still not clear what Windows users do

jmholzer replied:
You're right about this, I'm sorry I missed it. I updated both this command and the previous Windows one with PowerShell commands that I verified as working.

stichbury reviewed (approved):
Nice! Thanks for the updates. Just a couple of minor comments, but this looks great to me.

SajidAlamQB commented on Makefile:
```diff
@@ -62,6 +62,10 @@ uninstall-pre-commit:
 	@./tools/print_env.sh
 
+databricks-build:
+	python setup.py bdist_wheel
```
Could we not use `package` here?

SajidAlamQB replied:
Ah, right makes sense happy to leave it.

SajidAlamQB commented on docs/source/contribution/development_for_databricks.md:
````diff
@@ -0,0 +1,89 @@
+make databricks-build
+```
+
+You should see a stream of messages being written to your terminal. Behind the scenes, `build-databricks` does the following:
````
```suggestion
You should see a stream of messages being written to your terminal. Behind the scenes, `databricks-build` does the following:
```

SajidAlamQB commented on docs/source/contribution/development_for_databricks.md:
````diff
@@ -0,0 +1,89 @@
+
+```bash
+# Linux or macOS
+export DATABRICKS_CLUSTER_ID=<your-cluster-id>
````
Would this be persistent between different shells?

jmholzer replied:
Setting an environment variable in this way makes it available only in the shell process in which it is defined. If a user wants to make the environment variable persistent it's quite easy to google a solution for their particular shell.

SajidAlamQB commented on tools/databricks_build.py:
```diff
@@ -0,0 +1,97 @@
+    return the file with the highest version number.
+    """
+    dist_path = Path(__file__).resolve().parent.parent / "dist"
+    print(dist_path)
```
```suggestion
```

jmholzer replied:
Cheers 😄.

SajidAlamQB reviewed (commented):
This looks awesome @jmholzer! 🌟 Left some minor comments, and have tested it manually and it all seems to work. Haven't gone into the DataBricks API code that deeply but it all looks good and will look into it further during the session.

merelcht commented on docs/source/contribution/development_for_databricks.md:
````diff
@@ -0,0 +1,89 @@
+}
+```
+
+Any runs of a Kedro project on this cluster will now reflect your latest local changes to Kedro. You can now test your changes to Kedro by using your cluster to run a Kedro project.
````
I'm guessing that with any new change you'd have to run `make databricks-build` again?

jmholzer replied:
That's right, this is a one-off run. The alternative would be to create a process which watches `/dist`, using e.g. `dbx sync`.

merelcht replied:
I think this makes sense for now. A continues sync process might be handy if we're building the kedro-databricks plugin, but by that time there might already be another way to do that.

SajidAlamQB commented on tools/databricks_build.py:
```diff
@@ -0,0 +1,97 @@
+
+
+def _get_build_file_path() -> Path:
+    """Get the path of the whl file to install. If multiple whl files are found,
+    return the file with the highest version number.
+    """
+    dist_path = Path(__file__).resolve().parent.parent / "dist"
```
Maybe it would be better to clean `dist` after each run of `make databricks-build`
